airbyte-integrations/bases/base-typing-deduping-test/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/RecordDiffer.java

```diff
@@ -0,0 +1,386 @@
+                foundMismatch = true;
+              }
+            }
+            // ... and then check the actual raw data for any subfields that we weren't expecting.
+            LinkedHashMap<String, JsonNode> extraColumns = checkForExtraOrNonNullFields(expectedRawData, actualRawData);
+            if (extraColumns.size() > 0) {
+              for (Map.Entry<String, JsonNode> extraColumn : extraColumns.entrySet()) {
+                mismatchedRecordMessage += generateFieldError("_airbyte_data." + extraColumn.getKey(), null, extraColumn.getValue());
+                foundMismatch = true;
+              }
+            }
+          } else {
+            // For all other columns, we can just compare their values directly.
+            JsonNode expectedValue = expectedRecord.get(column);
+            JsonNode actualValue = actualRecord.get(column);
+            if (jsonNodesNotEquivalent(expectedValue, actualValue)) {
+              mismatchedRecordMessage += generateFieldError("column " + column, expectedValue, actualValue);
+              foundMismatch = true;
+            }
+          }
+        }
+        // Then check the entire actual record for any columns that we weren't expecting.
+        LinkedHashMap<String, JsonNode> extraColumns = checkForExtraOrNonNullFields(expectedRecord, actualRecord);
+        if (extraColumns.size() > 0) {
+          for (Map.Entry<String, JsonNode> extraColumn : extraColumns.entrySet()) {
+            mismatchedRecordMessage += generateFieldError("column " + extraColumn.getKey(), null, extraColumn.getValue());
+            foundMismatch = true;
+          }
+        }
+        if (foundMismatch) {
+          message += mismatchedRecordMessage;
+        }
+
+        expectedRecordIndex++;
+        actualRecordIndex++;
+      } else if (compare < 0) {
+        // The expected record is missing from the actual records. Print it and move on to the next expected
+        // record.
+        message += "Row was expected but missing: " + expectedRecord + "\n";
+        expectedRecordIndex++;
+      } else {
+        // There's an actual record which isn't present in the expected records. Print it and move on to the
+        // next actual record.
+        message += "Row was not expected but present: " + actualRecord + "\n";
+        actualRecordIndex++;
+      }
+    }
+    // Tail loops in case we reached the end of one list before the other.
+    while (expectedRecordIndex < expectedRecords.size()) {
+      message += "Row was expected but missing: " + expectedRecords.get(expectedRecordIndex) + "\n";
+      expectedRecordIndex++;
+    }
+    while (actualRecordIndex < actualRecords.size()) {
+      message += "Row was not expected but present: " + actualRecords.get(actualRecordIndex) + "\n";
+      actualRecordIndex++;
+    }
+
+    return message;
+  }
+
+  private static boolean jsonNodesNotEquivalent(JsonNode expectedValue, JsonNode actualValue) {
```

Comment:
This is ok but maybe the logic would look simpler to have a `jsonNodesAreEquivalent` method and this method just returns `!jsonNodesAreEquivalent(expectedValue, actualValue)`

Reply:
I flipped the method name + added negations at the if-statements https://github.com/airbytehq/airbyte/pull/27852/commits/6d437da026aff24db93ad94e043469746b0fb3c4. Normally I avoid methods returning negations, no idea why I wrote it this way to begin with

Reply:
... it would help if I actually changed the contents of the method to match its name though